File: days/2026-09-03/ai/linear_autograd/starter/src/linear_train.c
#include <stdio.h>

int main(void) {
    const double xs[] = {1.0, 2.0, 3.0, 4.0};
    const double ys[] = {3.0, 5.0, 7.0, 9.0};
    const int count = 4;

    double weight = 0.0;
    double bias = 0.0;
    const double learning_rate = 0.01;

    for (int epoch = 0; epoch < 1000; ++epoch) {
        double d_weight = 0.0;
        double d_bias = 0.0;

        for (int i = 0; i < count; ++i) {
            const double prediction = weight * xs[i] + bias;
            const double error = prediction - ys[i];

            // TODO: derive and accumulate dL/dw and dL/db.
            (void)error;
        }

        // TODO: average the gradients over the batch.
        // TODO: perform the SGD update.
        (void)d_weight;
        (void)d_bias;
        (void)learning_rate;
    }

    printf("w=%.6f b=%.6f\n", weight, bias);
    return 0;
}

File: days/2026-09-03/ai/linear_autograd/starter/python/linear_train.py
from __future__ import annotations


def train(epochs: int = 1000, learning_rate: float = 0.01) -> tuple[float, float]:
    xs = [1.0, 2.0, 3.0, 4.0]
    ys = [3.0, 5.0, 7.0, 9.0]

    weight = 0.0
    bias = 0.0

    for _ in range(epochs):
        d_weight = 0.0
        d_bias = 0.0

        for x, target in zip(xs, ys):
            prediction = weight * x + bias
            error = prediction - target

            # TODO [AI-PY-GRAD-01]: derive dL/dw and dL/db and accumulate them.
            _ = (x, error)

        # TODO [AI-PY-SGD-01]: average gradients and update weight/bias using SGD.
        _ = (d_weight, d_bias, learning_rate)

    return weight, bias


if __name__ == "__main__":
    w, b = train()
    print(f"w={w:.6f} b={b:.6f}")
